// discordx/build/esm/decorators/decorators/Once.js
import { DOn, MetadataStorage } from "../../index.js";
/**
 * Handle discord events once only with a defined handler
 *
 * @param options - Event parameters
 * ___
 *
 * [View Documentation](https://discordx.js.org/docs/discordx/decorators/general/once)
 *
 * @category Decorator
 */
export function Once(options) {
    return function (target, key, descriptor) {
        const clazz = target;
        const on = DOn.create({
            botIds: options?.botIds,
            event: options?.event ?? key,
            once: true,
            priority: options?.priority,
            rest: false,
        }).decorate(clazz.constructor, key, descriptor.value);
        MetadataStorage.instance.addOn(on);
    };
}
/**
 * Handle discord rest events with a defined handler
 *
 * @param options - Rest event options
 * ___
 *
 * [View Documentation](https://discordx.js.org/docs/discordx/decorators/general/once#rest)
 *
 * @category Decorator
 */
Once.rest = function (options) {
    return function (target, key, descriptor) {
        const clazz = target;
        const on = DOn.create({
            botIds: options?.botIds,
            event: options?.event ?? key,
            once: true,
            priority: options?.priority,
            rest: true,
        }).decorate(clazz.constructor, key, descriptor?.value);
        MetadataStorage.instance.addOn(on);
    };
};
