// discordx/build/esm/decorators/decorators/Slash.js
import { ApplicationCommandType } from "discord.js";
import { DApplicationCommand, MetadataStorage, SlashNameValidator, } from "../../index.js";
/**
 * Handle a slash command with a defined name
 *
 * @param options - Application command options
 * ___
 *
 * [View Documentation](https://discordx.js.org/docs/discordx/decorators/command/slash)
 *
 * @category Decorator
 */
export function Slash(options) {
    return function (target, key) {
        const name = options?.name ?? key;
        SlashNameValidator(name);
        const applicationCommand = DApplicationCommand.create({
            botIds: options?.botIds,
            defaultMemberPermissions: options?.defaultMemberPermissions,
            description: options?.description,
            descriptionLocalizations: options?.descriptionLocalizations,
            dmPermission: options?.dmPermission ?? true,
            guilds: options?.guilds,
            name: name,
            nameLocalizations: options?.nameLocalizations,
            type: ApplicationCommandType.ChatInput,
        }).decorate(target.constructor, key, target[key]);
        MetadataStorage.instance.addApplicationCommandSlash(applicationCommand);
    };
}
