// discordx/build/cjs/decorators/classes/DSimpleCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DSimpleCommand = void 0;
const discord_js_1 = require("discord.js");
const index_js_1 = require("../../index.js");
const Method_js_1 = require("./Method.js");
/**
 * @category Decorator
 */
class DSimpleCommand extends Method_js_1.Method {
    _description;
    _name;
    _prefix;
    _directMessage;
    _argSplitter;
    _options = [];
    _guilds;
    _botIds;
    _aliases;
    get aliases() {
        return this._aliases;
    }
    set aliases(value) {
        this._aliases = value;
    }
    get botIds() {
        return this._botIds;
    }
    set botIds(value) {
        this._botIds = value;
    }
    get prefix() {
        return this._prefix;
    }
    set prefix(value) {
        this._prefix = value;
    }
    get guilds() {
        return this._guilds;
    }
    set guilds(value) {
        this._guilds = value;
    }
    get argSplitter() {
        return this._argSplitter;
    }
    set argSplitter(value) {
        this._argSplitter = value;
    }
    get directMessage() {
        return this._directMessage;
    }
    set directMessage(value) {
        this._directMessage = value;
    }
    get name() {
        return this._name;
    }
    set name(value) {
        this._name = value;
    }
    get description() {
        return this._description;
    }
    set description(value) {
        this._description = value;
    }
    get options() {
        return this._options;
    }
    set options(value) {
        this._options = value;
    }
    constructor(data) {
        super();
        this._name = data.name;
        this._description = data.description ?? this.name;
        this._directMessage = data.directMessage ?? true;
        this._argSplitter = data.argSplitter;
        this._options = [];
        this._prefix = data.prefix;
        this._guilds = data.guilds ?? [];
        this._botIds = data.botIds ?? [];
        this._aliases = data.aliases ?? [];
    }
    static create(data) {
        return new DSimpleCommand(data);
    }
    isBotAllowed(botId) {
        if (!this.botIds.length) {
            return true;
        }
        return this.botIds.includes(botId);
    }
    async getGuilds(client, command) {
        const guilds = await (0, index_js_1.resolveIGuilds)(client, command, [
            ...client.botGuilds,
            ...this.guilds,
        ]);
        return guilds;
    }
    async isGuildAllowed(client, command, guildId) {
        if (!guildId) {
            return true;
        }
        const guilds = await this.getGuilds(client, command);
        if (!guilds.length) {
            return true;
        }
        return guilds.includes(guildId);
    }
    parseParams(command) {
        return command.options;
    }
    parseParamsEx(command) {
        if (!this.options.length) {
            return Promise.resolve([]);
        }
        const splitterEx = this.argSplitter ?? command.splitter ?? " ";
        const args = typeof splitterEx === "function"
            ? splitterEx(command)
            : command.argString
                .split(splitterEx)
                .filter((op) => op?.length)
                .map((op) => op.trim());
        return Promise.all(this.options
            .sort((a, b) => (a.index ?? 0) - (b.index ?? 0))
            .map((op, index) => {
            // only digits
            const id = args[index]?.replace(/\D/g, "");
            const invalidError = Error(`Invalid id given: ${args[index] ?? "unknown"}`);
            // undefined
            if (!args[index]?.length) {
                return undefined;
            }
            // Boolean
            if (op.type === index_js_1.SimpleCommandOptionType.Boolean) {
                const option = args[index];
                if (option === undefined) {
                    return undefined;
                }
                if (option.toLocaleLowerCase() === "false" ||
                    option.toLocaleLowerCase() === "0") {
                    return false;
                }
                return Boolean(option);
            }
            // Number
            if (op.type === index_js_1.SimpleCommandOptionType.Number) {
                return Number(args[index]);
            }
            // Channel | undefined
            if (op.type === index_js_1.SimpleCommandOptionType.Channel) {
                if (!id?.length || id.length < 16 || id.length > 20) {
                    return invalidError;
                }
                return command.message.guild?.channels
                    .fetch(id)
                    .catch((err) => err);
            }
            // Role | undefined
            if (op.type === index_js_1.SimpleCommandOptionType.Role) {
                if (!id?.length || id.length < 16 || id.length > 20) {
                    return invalidError;
                }
                return command.message.guild?.roles.fetch(id).catch((err) => err);
            }
            // GuildMember | User | undefined
            if (op.type === index_js_1.SimpleCommandOptionType.User) {
                if (!id?.length || id.length < 16 || id.length > 20) {
                    return invalidError;
                }
                if (command.message.channel.type === discord_js_1.ChannelType.DM) {
                    return command.message.client.user?.id === id
                        ? command.message.client.user
                        : command.message.author.id === id
                            ? command.message.author
                            : invalidError;
                }
                return command.message.guild?.members.fetch(id).catch((err) => err);
            }
            // GuildMember | User | Role | undefined
            if (op.type === index_js_1.SimpleCommandOptionType.Mentionable) {
                if (!id?.length || id.length < 16 || id.length > 20) {
                    return invalidError;
                }
                if (command.message.channel.type === discord_js_1.ChannelType.DM) {
                    return command.message.client.user?.id === id
                        ? command.message.client.user
                        : command.message.author.id === id
                            ? command.message.author
                            : invalidError;
                }
                return (command.message.guild?.members.fetch(id).catch((err) => err) ??
                    command.message.guild?.roles.fetch(id).catch((err) => err));
            }
            // string
            return args[index];
        }));
    }
}
exports.DSimpleCommand = DSimpleCommand;
