// discordx/build/esm/decorators/classes/DApplicationCommandOption.js
import { Decorator } from "@discordx/internal";
import { ApplicationCommandOptionType } from "discord.js";
/**
 * @category Decorator
 */
export class DApplicationCommandOption extends Decorator {
    _autocomplete;
    _channelTypes = undefined;
    _choices = [];
    _description;
    _descriptionLocalizations;
    _name;
    _nameLocalizations;
    _maxValue;
    _minValue;
    _maxLength;
    _minLength;
    _options = [];
    _required = true;
    _type;
    get autocomplete() {
        return this._autocomplete;
    }
    set autocomplete(value) {
        this._autocomplete = value;
    }
    get channelTypes() {
        return this._channelTypes;
    }
    set channelTypes(value) {
        this._channelTypes = value;
    }
    get choices() {
        return this._choices;
    }
    set choices(value) {
        this._choices = value;
    }
    get description() {
        return this._description;
    }
    set description(value) {
        this._description = value;
    }
    get descriptionLocalizations() {
        return this._descriptionLocalizations;
    }
    set descriptionLocalizations(value) {
        this._descriptionLocalizations = value;
    }
    get isNode() {
        return (this.type === ApplicationCommandOptionType.Subcommand ||
            this.type === ApplicationCommandOptionType.SubcommandGroup);
    }
    get maxValue() {
        return this._maxValue;
    }
    set maxValue(value) {
        this._maxValue = value;
    }
    get minValue() {
        return this._minValue;
    }
    set minValue(value) {
        this._minValue = value;
    }
    get maxLength() {
        return this._maxLength;
    }
    set maxLength(value) {
        this._maxLength = value;
    }
    get minLength() {
        return this._minLength;
    }
    set minLength(value) {
        this._minLength = value;
    }
    get name() {
        return this._name;
    }
    set name(value) {
        this._name = value;
    }
    get nameLocalizations() {
        return this._nameLocalizations;
    }
    set nameLocalizations(value) {
        this._nameLocalizations = value;
    }
    get options() {
        return this._options;
    }
    set options(value) {
        this._options = value;
    }
    get required() {
        return this._required;
    }
    set required(value) {
        this._required = value;
    }
    get type() {
        return this._type;
    }
    set type(value) {
        this._type = value;
    }
    constructor(data) {
        super();
        this._name = data.name;
        this._autocomplete = data.autocomplete;
        this._channelTypes = data.channelType?.sort();
        this._description = data.description;
        this._index = data.index;
        this._maxValue = data.maxValue;
        this._minValue = data.minValue;
        this._maxLength = data.maxLength;
        this._minLength = data.minLength;
        this._required = data.required ?? false;
        this._type = data.type;
        this._descriptionLocalizations = data.descriptionLocalizations ?? null;
        this._nameLocalizations = data.nameLocalizations ?? null;
    }
    static create(data) {
        return new DApplicationCommandOption(data);
    }
    toJSON() {
        const options = [...this.options]
            .reverse()
            .map((option) => option.toJSON());
        const data = {
            autocomplete: this.autocomplete ? true : undefined,
            channelTypes: this.channelTypes,
            choices: this.isNode
                ? undefined
                : this.choices.length === 0
                    ? undefined
                    : this.choices.map((choice) => choice.toJSON()),
            description: this.description,
            descriptionLocalizations: this.descriptionLocalizations,
            maxLength: this.maxLength,
            maxValue: this.maxValue,
            minLength: this.minLength,
            minValue: this.minValue,
            name: this.name,
            nameLocalizations: this.nameLocalizations,
            options: options.length === 0 ? undefined : options,
            required: this.isNode ? undefined : this.required,
            type: this.type,
        };
        return data;
    }
}
