// discordx/build/esm/decorators/classes/DComponent.js
import { resolveIGuilds } from "../../index.js";
import { Method } from "./Method.js";
/**
 * @category Decorator
 */
export class DComponent extends Method {
    _type;
    _id;
    _guilds;
    _botIds;
    get type() {
        return this._type;
    }
    get botIds() {
        return this._botIds;
    }
    set botIds(value) {
        this._botIds = value;
    }
    get id() {
        return this._id;
    }
    set id(value) {
        this._id = value;
    }
    get guilds() {
        return this._guilds;
    }
    set guilds(value) {
        this._guilds = value;
    }
    constructor(data) {
        super();
        this._type = data.type;
        this._id = data.id;
        this._guilds = data.guilds ?? [];
        this._botIds = data.botIds ?? [];
    }
    static create(data) {
        return new DComponent(data);
    }
    isBotAllowed(botId) {
        if (!this.botIds.length) {
            return true;
        }
        return this.botIds.includes(botId);
    }
    async getGuilds(client) {
        const guilds = await resolveIGuilds(client, this, [
            ...client.botGuilds,
            ...this.guilds,
        ]);
        return guilds;
    }
    async isGuildAllowed(client, guildId) {
        if (!guildId) {
            return true;
        }
        const guilds = await this.getGuilds(client);
        if (!guilds.length) {
            return true;
        }
        return guilds.includes(guildId);
    }
    isId(text) {
        return typeof this.id === "string" ? this.id === text : this.id.test(text);
    }
    parseParams() {
        return [];
    }
}
