// discordx/build/esm/decorators/classes/DOn.js
import { Method } from "./Method.js";
/**
 * @category Decorator
 */
export class DOn extends Method {
    _event;
    _once;
    _rest;
    _priority;
    _botIds;
    get botIds() {
        return this._botIds;
    }
    set botIds(value) {
        this._botIds = value;
    }
    get event() {
        return this._event;
    }
    set event(value) {
        this._event = value;
    }
    get once() {
        return this._once;
    }
    set once(value) {
        this._once = value;
    }
    get priority() {
        return this._priority;
    }
    set priority(value) {
        this._priority = value;
    }
    get rest() {
        return this._rest;
    }
    set rest(value) {
        this._rest = value;
    }
    constructor(data) {
        super();
        this._event = data.event;
        this._once = data.once;
        this._priority = data.priority ?? Number.MAX_SAFE_INTEGER;
        this._rest = data.rest;
        this._botIds = data.botIds ?? [];
    }
    static create(data) {
        return new DOn(data);
    }
    isBotAllowed(botId) {
        if (!this.botIds.length) {
            return true;
        }
        return this.botIds.includes(botId);
    }
    parseParams() {
        return [];
    }
}
