// discordx/build/cjs/decorators/decorators/ContextMenu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextMenu = void 0;
const discord_js_1 = require("discord.js");
const index_js_1 = require("../../index.js");
/**
 * Interact with context menu with a defined identifier
 *
 * @param options - Application command options
 * ___
 *
 * [View Documentation](https://discordx.js.org/docs/discordx/decorators/gui/context-menu)
 *
 * @category Decorator
 */
function ContextMenu(options) {
    return function (target, key) {
        const applicationCommand = index_js_1.DApplicationCommand.create({
            botIds: options.botIds,
            defaultMemberPermissions: options.defaultMemberPermissions,
            description: "",
            dmPermission: options.dmPermission,
            guilds: options.guilds,
            name: options.name ?? key,
            nameLocalizations: options.nameLocalizations,
            type: options.type,
        }).decorate(target.constructor, key, target[key]);
        if (options.type === discord_js_1.ApplicationCommandType.Message) {
            index_js_1.MetadataStorage.instance.addApplicationCommandMessage(applicationCommand);
        }
        else {
            index_js_1.MetadataStorage.instance.addApplicationCommandUser(applicationCommand);
        }
    };
}
exports.ContextMenu = ContextMenu;
