// discordx/build/cjs/decorators/decorators/On.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.On = void 0;
const index_js_1 = require("../../index.js");
/**
 * Handle discord events with a defined handler
 *
 * @param options - Event options
 * ___
 *
 * [View Documentation](https://discordx.js.org/docs/discordx/decorators/general/on)
 *
 * @category Decorator
 */
function On(options) {
    return function (target, key, descriptor) {
        const clazz = target;
        const on = index_js_1.DOn.create({
            botIds: options?.botIds,
            event: options?.event ?? key,
            once: false,
            priority: options?.priority,
            rest: false,
        }).decorate(clazz.constructor, key, descriptor?.value);
        index_js_1.MetadataStorage.instance.addOn(on);
    };
}
exports.On = On;
/**
 * Handle discord rest events with a defined handler
 *
 * @param options - Rest event options
 * ___
 *
 * [View Documentation](https://discordx.js.org/docs/discordx/decorators/general/on#rest)
 *
 * @category Decorator
 */
On.rest = function (options) {
    return function (target, key, descriptor) {
        const clazz = target;
        const on = index_js_1.DOn.create({
            botIds: options?.botIds,
            event: options?.event ?? key,
            once: false,
            priority: options?.priority,
            rest: true,
        }).decorate(clazz.constructor, key, descriptor?.value);
        index_js_1.MetadataStorage.instance.addOn(on);
    };
};
