// discordx/build/cjs/decorators/decorators/Once.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Once = void 0;
const index_js_1 = require("../../index.js");
/**
 * Handle discord events once only with a defined handler
 *
 * @param options - Event parameters
 * ___
 *
 * [View Documentation](https://discordx.js.org/docs/discordx/decorators/general/once)
 *
 * @category Decorator
 */
function Once(options) {
    return function (target, key, descriptor) {
        const clazz = target;
        const on = index_js_1.DOn.create({
            botIds: options?.botIds,
            event: options?.event ?? key,
            once: true,
            priority: options?.priority,
            rest: false,
        }).decorate(clazz.constructor, key, descriptor.value);
        index_js_1.MetadataStorage.instance.addOn(on);
    };
}
exports.Once = Once;
/**
 * Handle discord rest events with a defined handler
 *
 * @param options - Rest event options
 * ___
 *
 * [View Documentation](https://discordx.js.org/docs/discordx/decorators/general/once#rest)
 *
 * @category Decorator
 */
Once.rest = function (options) {
    return function (target, key, descriptor) {
        const clazz = target;
        const on = index_js_1.DOn.create({
            botIds: options?.botIds,
            event: options?.event ?? key,
            once: true,
            priority: options?.priority,
            rest: true,
        }).decorate(clazz.constructor, key, descriptor?.value);
        index_js_1.MetadataStorage.instance.addOn(on);
    };
};
