// discordx/build/cjs/decorators/decorators/SimpleCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleCommand = void 0;
const index_js_1 = require("../../index.js");
/**
 * Handle a simple command with a defined name
 *
 * Example ``!hello world``
 *
 * @param options - Command options
 * ___
 *
 * [View Documentation](https://discordx.js.org/docs/discordx/decorators/command/simple-command)
 *
 * @category Decorator
 */
function SimpleCommand(options) {
    return function (target, key) {
        const cmd = index_js_1.DSimpleCommand.create({
            aliases: options?.aliases,
            argSplitter: options?.argSplitter,
            botIds: options?.botIds,
            description: options?.description,
            directMessage: options?.directMessage,
            guilds: options?.guilds,
            name: options?.name ?? key,
            prefix: options?.prefix,
        }).decorate(target.constructor, key, target[key]);
        index_js_1.MetadataStorage.instance.addSimpleCommand(cmd);
    };
}
exports.SimpleCommand = SimpleCommand;
