// discordx/build/cjs/decorators/decorators/Slash.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Slash = void 0;
const discord_js_1 = require("discord.js");
const index_js_1 = require("../../index.js");
/**
 * Handle a slash command with a defined name
 *
 * @param options - Application command options
 * ___
 *
 * [View Documentation](https://discordx.js.org/docs/discordx/decorators/command/slash)
 *
 * @category Decorator
 */
function Slash(options) {
    return function (target, key) {
        const name = options?.name ?? key;
        (0, index_js_1.SlashNameValidator)(name);
        const applicationCommand = index_js_1.DApplicationCommand.create({
            botIds: options?.botIds,
            defaultMemberPermissions: options?.defaultMemberPermissions,
            description: options?.description,
            descriptionLocalizations: options?.descriptionLocalizations,
            dmPermission: options?.dmPermission ?? true,
            guilds: options?.guilds,
            name: name,
            nameLocalizations: options?.nameLocalizations,
            type: discord_js_1.ApplicationCommandType.ChatInput,
        }).decorate(target.constructor, key, target[key]);
        index_js_1.MetadataStorage.instance.addApplicationCommandSlash(applicationCommand);
    };
}
exports.Slash = Slash;
